### PR TITLE
StreamInput throws exceptions instead of using assertions

### PR DESCRIPTION
**Pull request by jaymode**
StreamInput has methods such as readVInt that perform sanity checks on the data using assertions,
which will catch bad data in tests but provide no safety when running as a node without assertions
enabled. The use of assertions also make testing with invalid data difficult since we would need
to handle assertion errors in the code using the stream input and errors like this should not be
something we try to catch. This commit changes StreamInput to throw an IOException instead of
using an assertion and also fixes a bug where we created an AssertionError and never did
anything with it.

### CODE REVIEW

**Comment by rjernst** [comment]
I don't think we should have a flag here. It would mean we are paying the cost of an additional branch all the time. The reason for an assertion is reading vbyte is usually done in a tight loop. If we want exceptions all the time here, let's check that hotspot is able to optimize away this check in the normal case by doing a perf test?

**Comment by rjernst** [comment]
And to be a little more clear, I don't think we can ever optimize away this completely, so I don't think it is a worthwhile change.  If there is a bug in code using readVByte, then it seems to be lacking tests, which would have caught the problem during testing instead of production.

**Comment by jaymode** [comment]
@rjernst I pushed a JMH benchmark. I ran tests and there is a slight impact, but I need to do some more runs:

Master:
```
Benchmark                                (size)  Mode  Cnt    Score    Error  Units
StreamInputBenchmark.testReadToArray    1048576  avgt   30    4.577 ±  0.093  ms/op
StreamInputBenchmark.testReadToArray   10485760  avgt   30   44.775 ±  0.474  ms/op
StreamInputBenchmark.testReadToArray  104857600  avgt   30  504.517 ± 35.405  ms/op
StreamInputBenchmark.testReadVInt       1048576  avgt   30    4.319 ±  0.048  ms/op
StreamInputBenchmark.testReadVInt      10485760  avgt   30   43.601 ±  0.511  ms/op
StreamInputBenchmark.testReadVInt     104857600  avgt   30  434.953 ±  5.873  ms/op
```

This branch:
```
Benchmark                                (size)  Mode  Cnt    Score    Error  Units
StreamInputBenchmark.testReadToArray    1048576  avgt   30    4.516 ±  0.042  ms/op
StreamInputBenchmark.testReadToArray   10485760  avgt   30   45.346 ±  0.424  ms/op
StreamInputBenchmark.testReadToArray  104857600  avgt   30  514.794 ± 36.346  ms/op
StreamInputBenchmark.testReadVInt       1048576  avgt   30    4.420 ±  0.084  ms/op
StreamInputBenchmark.testReadVInt      10485760  avgt   30   44.364 ±  0.715  ms/op
StreamInputBenchmark.testReadVInt     104857600  avgt   30  439.756 ±  7.671  ms/op
```

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 1]
```diff
@@ -202,7 +220,13 @@ public int readVInt() throws IOException {
             return i;
         }
         b = readByte();
-        assert (b & 0x80) == 0;
+        if (useExceptions) {
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: I think we can drop this flag and just explode if the stream is broken.

**Review by jasontedor — COMMENTED** [review]
I chatted with @jaymode via another channel and I left a comment based on our discussion.

File: benchmarks/src/main/java/org/elasticsearch/benchmark/stream/StreamInputBenchmark.java [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+import java.util.concurrent.TimeUnit;
+
+@Fork(10)
+@Warmup(iterations = 10)
+@Measurement(iterations = 10)
+@BenchmarkMode(Mode.AverageTime)
+@OutputTimeUnit(TimeUnit.NANOSECONDS)
+@State(Scope.Benchmark)
+@SuppressWarnings("unused") //invoked by benchmarking framework
+public class StreamInputBenchmark {
```
Comment (jasontedor) [inline benchmarks/src/main/java/org/elasticsearch/benchmark/stream/StreamInputBenchmark.java]: I think we can drop the benchmark.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 3]
```diff
@@ -367,7 +369,7 @@ public String readString() throws IOException {
                     buffer[i] = ((char) ((c & 0x0F) << 12 | (readByte() & 0x3F) << 6 | (readByte() & 0x3F) << 0));
                     break;
                 default:
-                    new AssertionError("unexpected character: " + c + " hex: " + Integer.toHexString(c));
+                    throw new IOException("Invalid string; unexpected character: " + c + " hex: " + Integer.toHexString(c));
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: Oh boy, I hope nothing is broken here.

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java [hunk 4]
```diff
@@ -202,7 +202,9 @@ public int readVInt() throws IOException {
             return i;
         }
         b = readByte();
-        assert (b & 0x80) == 0;
+        if ((b & 0x80) != 0) {
+            throw new IOException("Invalid vInt " + Integer.toHexString(b) + " & 0x80 = " + Integer.toHexString((b & 0x80)));
```
Comment (jasontedor) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: `Integer.toHexString(b & 0x80):` Isn't this going to be `0x80` 100% of the time? that we are here? I think maybe we should output the current value of the accumulated bytes though (`i`)?

**Review by jasontedor — COMMENTED** [review]
I left some comments.

File: benchmarks/src/main/java/org/elasticsearch/benchmark/stream/StreamInputBenchmark.java (hunk 2 above)
Comment (jaymode) [inline benchmarks/src/main/java/org/elasticsearch/benchmark/stream/StreamInputBenchmark.java]: Will drop it. I did however run this benchmark once with the version you reviewed.

Master:

```
On master:
# Run complete. Total time: 01:02:39

Benchmark                                      (size)  Mode  Cnt          Score         Error  Units
StreamInputBenchmark.measureReadAndSumVInt          4  avgt  100         35.233 ±       0.244  ns/op
StreamInputBenchmark.measureReadAndSumVInt         16  avgt  100         85.349 ±       1.447  ns/op
StreamInputBenchmark.measureReadAndSumVInt         64  avgt  100        235.382 ±       2.569  ns/op
StreamInputBenchmark.measureReadAndSumVInt        256  avgt  100        902.952 ±       9.297  ns/op
StreamInputBenchmark.measureReadAndSumVInt       1024  avgt  100       3499.609 ±      38.687  ns/op
StreamInputBenchmark.measureReadAndSumVInt      16348  avgt  100      54520.067 ±     371.619  ns/op
StreamInputBenchmark.measureReadAndSumVInt    1048576  avgt  100    3374447.046 ±   33979.659  ns/op
StreamInputBenchmark.measureReadAndSumVInt   10485760  avgt  100   34004175.888 ±  170539.485  ns/op
StreamInputBenchmark.measureReadAndSumVInt  104857600  avgt  100  344504176.762 ± 4489346.992  ns/op
StreamInputBenchmark.measureReadToArray             4  avgt  100         44.798 ±       0.536  ns/op
StreamInputBenchmark.measureReadToArray            16  avgt  100         96.208 ±       1.272  ns/op
StreamInputBenchmark.measureReadToArray            64  avgt  100        252.820 ±       3.105  ns/op
StreamInputBenchmark.measureReadToArray           256  avgt  100        920.067 ±       6.403  ns/op
StreamInputBenchmark.measureReadToArray          1024  avgt  100       3587.619 ±      31.852  ns/op
StreamInputBenchmark.measureReadToArray         16348  avgt  100      56473.895 ±     386.739  ns/op
StreamInputBenchmark.measureReadToArray       1048576  avgt  100    3475751.564 ±   45444.250  ns/op
StreamInputBenchmark.measureReadToArray      10485760  avgt  100   34633952.391 ±  363944.894  ns/op
StreamInputBenchmark.measureReadToArray     104857600  avgt  100  347333659.817 ± 3212549.852  ns/op
```

This branch:

```
# Run complete. Total time: 01:02:34

Benchmark                                      (size)  Mode  Cnt          Score         Error  Units
StreamInputBenchmark.measureReadAndSumVInt          4  avgt  100         36.267 ±       0.785  ns/op
StreamInputBenchmark.measureReadAndSumVInt         16  avgt  100         85.695 ±       0.678  ns/op
StreamInputBenchmark.measureReadAndSumVInt         64  avgt  100        236.789 ±       1.683  ns/op
StreamInputBenchmark.measureReadAndSumVInt        256  avgt  100        899.875 ±       9.242  ns/op
StreamInputBenchmark.measureReadAndSumVInt       1024  avgt  100       3502.975 ±      31.820  ns/op
StreamInputBenchmark.measureReadAndSumVInt      16348  avgt  100      55528.773 ±     286.482  ns/op
StreamInputBenchmark.measureReadAndSumVInt    1048576  avgt  100    3432045.939 ±   23759.513  ns/op
StreamInputBenchmark.measureReadAndSumVInt   10485760  avgt  100   34495218.291 ±  179885.969  ns/op
StreamInputBenchmark.measureReadAndSumVInt  104857600  avgt  100  349996045.803 ± 1013571.648  ns/op
StreamInputBenchmark.measureReadToArray             4  avgt  100         45.405 ±       0.918  ns/op
StreamInputBenchmark.measureReadToArray            16  avgt  100         96.805 ±       0.988  ns/op
StreamInputBenchmark.measureReadToArray            64  avgt  100        254.719 ±       2.286  ns/op
StreamInputBenchmark.measureReadToArray           256  avgt  100        937.991 ±       7.558  ns/op
StreamInputBenchmark.measureReadToArray          1024  avgt  100       3656.997 ±      29.638  ns/op
StreamInputBenchmark.measureReadToArray         16348  avgt  100      57338.926 ±     292.516  ns/op
StreamInputBenchmark.measureReadToArray       1048576  avgt  100    3497268.663 ±   24068.370  ns/op
StreamInputBenchmark.measureReadToArray      10485760  avgt  100   35388540.405 ±  447633.734  ns/op
StreamInputBenchmark.measureReadToArray     104857600  avgt  100  357208368.590 ± 3177542.818  ns/op
```

File: core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java (hunk 4 above)
Comment (jaymode) [inline core/src/main/java/org/elasticsearch/common/io/stream/StreamInput.java]: You're correct; I updated it to match the format used in readVLong so it should be more useful

**Review by rjernst — APPROVED** [review]
LGTM. Thanks for the perf testing. Looking at it now, this is both an extreme case (to see 4 bytes) so the condition would be checked very rarely, and we also do not do streams in ES in tight loops as we do with decoding posting lists in lucene with vbyte (which was my initial thought).

**Review by jasontedor — APPROVED** [review]
LGTM.

**Comment by jaymode** [comment]
Thanks @rjernst and @jasontedor for the reviews

**Pull request merged**